video/out/vulkan/context_android.c

```diff
@@ -0,0 +1,128 @@
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <vulkan/vulkan.h>
+#include <vulkan/vulkan_android.h>
+#include <libavcodec/jni.h>
+#include <android/native_window_jni.h>
+
+#include "common.h"
+#include "options/m_config.h"
+#include "context.h"
+#include "utils.h"
+
+struct android_opts {
+    struct m_geometry surface_size;
+};
+
+extern const struct m_sub_options android_conf; // lol
+
+struct priv {
+    struct mpvk_ctx vk;
+    ANativeWindow *native_window;
+};
+
+static void android_uninit(struct ra_ctx *ctx)
+{
+    struct priv *p = ctx->priv;
+    ra_vk_ctx_uninit(ctx);
+    mpvk_uninit(&p->vk);
+
+    if (p->native_window) {
+        ANativeWindow_release(p->native_window);
+        p->native_window = NULL;
+    }
+}
+
+static bool android_init(struct ra_ctx *ctx)
+{
+    struct priv *p = ctx->priv = talloc_zero(ctx, struct priv);
+    struct mpvk_ctx *vk = &p->vk;
+    int msgl = ctx->opts.probing ? MSGL_V : MSGL_ERR;
+
+    if (!mpvk_init(vk, ctx, VK_KHR_ANDROID_SURFACE_EXTENSION_NAME))
+        goto fail;
+
+    jobject surface = (jobject)(intptr_t)ctx->vo->opts->WinID;
+    JavaVM *vm = (JavaVM *)av_jni_get_java_vm(NULL);
+    JNIEnv *env;
+    int ret = (*vm)->GetEnv(vm, (void**)&env, JNI_VERSION_1_6);
+    if (ret == JNI_EDETACHED) {
+        if ((*vm)->AttachCurrentThread(vm, &env, NULL) != 0) {
+            MP_FATAL(ctx, "Could not attach java VM.\n");
+            goto fail;
+        }
+    }
+    p->native_window = ANativeWindow_fromSurface(env, surface);
+    (*vm)->DetachCurrentThread(vm);
+
+    VkAndroidSurfaceCreateInfoKHR info = {
+         .sType = VK_STRUCTURE_TYPE_ANDROID_SURFACE_CREATE_INFO_KHR,
+         .window = p->native_window
+    };
+
+    VkInstance inst = vk->vkinst->instance;
+    VkResult res = vkCreateAndroidSurfaceKHR(inst, &info, NULL, &vk->surface);
+    if (res != VK_SUCCESS) {
+        MP_MSG(ctx, msgl, "Failed creating Android surface\n");
+        goto fail;
+    }
+
+    if (!ra_vk_ctx_init(ctx, vk, VK_PRESENT_MODE_FIFO_KHR))
+        goto fail;
+
+    return true;
+fail:
+    android_uninit(ctx);
+    return false;
+}
+
+static bool android_reconfig(struct ra_ctx *ctx)
+{
+    struct priv *p = ctx->priv;
+    void *tmp = talloc_new(NULL);
+    struct android_opts *opts = mp_get_config_group(tmp, ctx->global, &android_conf);
+    int w = opts->surface_size.w, h = opts->surface_size.h;
+
+    if (!w)
+        w = ANativeWindow_getWidth(p->native_window);
+    if (!h)
+        h = ANativeWindow_getHeight(p->native_window);
+
+    if (!w || !h) {
+        MP_FATAL(ctx, "Failed to get height and width!\n");
+        return false;
```

Comment:
I've skimmed over talloc for only like 5 minutes so I may be completely wrong here but doesn't this need a `talloc_free(tmp)` before the return or we leak some memory?

Reply:
Yeah, looks like a memory leak.

In fact, since this is copied from the existing context code the same memory leak exists there:
https://github.com/mpv-player/mpv/blob/61b46d101e95ce29c680082cc11ae72c7b80e045/video/out/opengl/context_android.c#L150-L158